File: design-pattern/02-adapter-pattern/adapterPatter.cpp
#include <iostream>

class Target
{
public:
    virtual void request() = 0;
};

class Adaptee
{
public:
    void specificRequest()
    {
        std::cout << "既存クラスのメソッドが呼ばれました。" << std::endl;
    }
};

class Adapter : public Target
{
private:
    Adaptee *adaptee;

public:
    Adapter(Adaptee *adaptee)
    {
        this->adaptee = adaptee;
    }

    void request()
    {
        adaptee->specificRequest();
    }
};

int main()
{
    Adaptee *adaptee = new Adaptee();
    Target *adapter = new Adapter(adaptee);
    adapter->request();
    return 0;
}

File: design-pattern/02-adapter-pattern/adapterPatter.py
class Target:
    """
    ターゲットクラス/インターフェース(抽象)
    """

    def request(self):
        pass


class Adaptee:
    """
    提供される既存クラス(具象)
    """

    def specific_request(self):
        print("既存クラスのメソッドが呼ばれました。")


class Adapter(Target):
    """
    アダプタークラス(具象)
    - Targetインタフェースを実装する
    - Adapteeクラスを覆う
    """

    def __init__(self, adaptee: Adaptee):
        # コンストラクタで、既存クラスのインスタンスを受け取る
        self.adaptee = adaptee

    def request(self):
        self.adaptee.specific_request()


if __name__ == '__main__':
    """
    Targetのメソッドを使うクライアント
    """
    adapter = Adapter(Adaptee())
    adapter.request()
